app/js/DocArchiveManager.js

```diff
@@ -1,295 +1,179 @@
+    lines: doc.lines,
+    ranges: doc.ranges,
+    schema_v: 1
+  })
 
-  _mongoDocToS3Doc(doc, callback) {
-    if (callback == null) {
-      callback = function (error, s3_doc) {}
-    }
-    if (doc.lines == null) {
-      return callback(new Error('doc has no lines'))
-    }
-    const json = JSON.stringify({
-      lines: doc.lines,
-      ranges: doc.ranges,
-      schema_v: 1
+  // this should never happen, but protects against memory-corruption errors that
+  // have happened in the past
+  if (json.indexOf('\u0000') > -1) {
+    const error = new Error('null bytes detected')
+    logger.err({ err: error, doc }, error.message)
+    throw error
+  }
+
+  const md5 = crypto.createHash('md5').update(json).digest('hex')
+  const stream = Streamifier.createReadStream(json)
+  await PersistorManager.sendStream(settings.docstore.bucket, key, stream, {
+    sourceMd5: md5
+  })
+  await promisify(MongoManager.markDocAsArchived)(doc._id, doc.rev)
+}
+
+async function unArchiveAllDocs(projectId) {
+  const docs = await promisify(MongoManager.getArchivedProjectDocs)(projectId)
+  if (!docs) {
+    throw new Errors.NotFoundError(`No docs for project ${projectId}`)
+  }
+  await AsyncPool(PARALLEL_JOBS, docs, (doc) =>
+    unarchiveDoc(projectId, doc._id)
+  )
+}
+
+async function unarchiveDoc(projectId, docId) {
+  logger.log(
+    { project_id: projectId, doc_id: docId },
+    'getting doc from persistor'
+  )
+  const key = `${projectId}/${docId}`
+  const sourceMd5 = await PersistorManager.getObjectMd5Hash(
+    settings.docstore.bucket,
+    key
+  )
+  const stream = await PersistorManager.getObjectStream(
+    settings.docstore.bucket,
+    key
+  )
+  stream.resume()
+  const json = await _streamToString(stream)
+  const md5 = crypto.createHash('md5').update(json).digest('hex')
+  if (sourceMd5 !== md5) {
+    throw new Errors.Md5MismatchError('md5 mismatch when downloading doc', {
+      key,
+      sourceMd5,
+      md5
     })
-    if (json.indexOf('\u0000') !== -1) {
-      const error = new Error('null bytes detected')
-      logger.err({ err: error, doc, json }, error.message)
-      return callback(error)
-    }
-    return callback(null, json)
-  },
+  }
 
-  buildS3Options(key) {
-    if (settings.docstore.s3 == null) {
-      throw new Error('S3 settings are not configured')
-    }
-    return {
-      aws: {
-        key: settings.docstore.s3.key,
-        secret: settings.docstore.s3.secret,
-        bucket: settings.docstore.s3.bucket
-      },
-      timeout: thirtySeconds,
-      uri: `https://${settings.docstore.s3.bucket}.s3.amazonaws.com/${key}`
+  const doc = JSON.parse(json)
+
+  const mongoDoc = {}
+  if (doc.schema_v === 1 && doc.lines != null) {
+    mongoDoc.lines = doc.lines
+    if (doc.ranges != null) {
+      mongoDoc.ranges = RangeManager.jsonRangesToMongo(doc.ranges)
     }
+  } else if (Array.isArray(doc)) {
+    mongoDoc.lines = doc
+  } else {
+    throw new Error("I don't understand the doc format in s3")
```

Comment:
Thought: are we using OError in this project? Would we be able to debug this issue if if popped up in production?

Reply:
I don't think we were, but it's being pulled in via `object-persistor` so I think it would be better to use it, yes. :+1: